README.md
# nvoip-node-sms

Send SMS messages in node via Nvoip SMS API.

## Installation

Install nvoip-node-sms from npm

with yarn:

~~~
yarn add nvoip-node-sms
~~~

or alternatively using npm:

~~~
npm install --save nvoip-node-sms
~~~

## Testing

With yarn:

~~~
yarn test
~~~

or alternatively using npm:

~~~
npm run test
~~~

## Basic usage

To be able to send messages, you'll need an access token provided by Nvoip. You can get it by signing in on https://www.nvoip.com.br. After signing in, go to 'Configurações' > 'Integrações' and click the button 'Abrir' on the 'ExactSales' div to launch a pop up window. You should be able to copy the token string (the string that follows the word 'Token') from there.

~~~javascript

// Import the client factory function:

const { createClient } = require('nvoip-node-sms');

// or alternatively using ES6 imports:
// import { createClient } from 'nvoip-node-sms';

// Instantiate a client:

const client = createClient({ authToken: '523abc21d8b659bc19553d5ce5b42' });

// Send a message!

client.sendSms('1111111111', 'Hi!')
  .then(response => console.log(response))
  .catch(error => console.error(error));

// That's it!
~~~

## Command line utility

nvoip-node-sms bundles with a simple command line utililty that enables you to send messages without the need to write any code.

You can check the available options using

~~~
./bin/send-sms --help
~~~

Be sure the script has the necessary permissions to execute. You can, of course, use `chmod +x ./bin/send-sms` for that.

Its also possible to run it without installation using npx.

## API

### `createClient(config)`

Creates a client for sending messages.

#### Parameters:

- config Object» a configuration object:
  - authToken «String» the authentication token.
  - \[url] «String» the API endpoint.
  - \[request] «Function» custom request function. (see the request module for more details).

#### Returns

A SMS client «SmsClient».

#### Example

~~~javascript

const client = createClient({
  url: 'https://api.nvoip.com.br/v1/sms',
  authToken: '523abc21d8b659bc19553d5ce5b42',
});

~~~

### `SmsClient.prototype.sendSms(phoneNumber, textMessage)`

- phoneNumber «String» a phone number containing nothing but digits.
- textMessage «String» the message to be sent.

#### Returns

A promise «Promise» containing the API response data.

#### Throws

- «MessageValidationError» if the textMessage is invalid.
- «PhoneNumberValidationError» if the phoneNumber is invalid.
- «HttpRequestError» if the HTTP request failed.

#### Example

~~~javascript

const client = createClient({
  authToken: '523abc21d8b659bc19553d5ce5b42',
});

client.sendSms('121212121', 'How are you doing?')
  .then(data => console.log(data));

~~~

### Error Handling

A collection of errors are provided in order to enable proper error handling and control flow. You can safely use the `instanceof` operator in a catch block to detemine the precise source and cause of the error.

Error hierarchy:

- «SmsClientError»
  - «HttpRequestError»
  - «ValidationError»
    - «PhoneNumberValidationError»
    - «MessageValidationError»

Check the errors module for more details.

Example:

~~~javascript

//...
const {
  ValidationError,
  HttpRequestError,
  createClient,
} = require('nvoip-node-sms');

// Immediately invoked asynchronous function
(async () => {

  // ...

  const client = createClient({
    authToken: '523abc21d8b659bc19553d5ce5b42',
  });

  // ...

  let response;

  try {
    response = await client.sendSms('1125235235', 'Hi!');
  } catch (err) {
    if (err instanceof ValidationError) {
      console.error('error: be more careful with your input!');
    } else if (err instanceof HttpRequestError) {
      console.error('error: maybe check your network connection?!');
    } else {
      // Unknown error, don't treat it:
      throw err;
    }
  }

  console.log(response);

  // ...
}())

// ...

~~~


CHANGELOG.md
# CHANGELOG

## 1.1.0

- Remove invalid characters check in text message.
- Change maximum message size from 140 to 160.
- Add API integration tests.
  - Add new command test-integrations.
  - Add .env.test.example file to set the test environment.

src/__tests__/fixtures.js
/**
 * A 160 character long text.
 * 
 * Obs.: there is a newline char to account at the end of every line.
 */
const TEXT_160_CHARS =`
01-qwerty
02-qwerty
03-qwerty
04-qwerty
05-qwerty
06-qwerty
07-qwerty
08-qwerty
09-qwerty
10-qwerty
11-qwerty
12-qwerty
13-qwerty
14-qwerty
15-qwerty
16-qwerty`;

/**
 * A 161 character long text.
 */
const TEXT_161_CHARS = TEXT_160_CHARS + '!';

module.exports = {
  TEXT_160_CHARS,
  TEXT_161_CHARS,
};


src/__tests__/index.test.js
const { createClient } = require('../index');

describe('createClient', () => {

  it('creates a SmsClient', async () => {
    const requestMock = jest.fn();
    const config = {
      url: 'https://github.com/fhpriamo',
      authToken: 'authtoken',
      request: requestMock,
    };

    const client = createClient(config);

    await client.sendSms('11111111111', ' ');

    expect(requestMock).toHaveBeenCalledWith(expect.objectContaining({
      url: config.url,
      headers: expect.objectContaining({
        'token_auth': config.authToken,
      }),
    }));
  });
});


src/__tests__/client.test.js
const { PhoneNumber, Message, SmsClient } = require('../client');
const { TEXT_160_CHARS, TEXT_161_CHARS } = require('./fixtures');
const {
  PhoneNumberValidationError,
  MessageValidationError,
  HttpRequestError,
} = require('../errors');

beforeEach(() => jest.clearAllMocks());

describe('client', () => {

  describe('PhoneNumber', () => {

    it.each`
      invalidPhone
      ${''},
      ${'ab888592835'},
      ${'32 92222 2835'},
      ${'32 92222-2835'},
      ${'(32) 922222835'},
      ${'(32)922222835'},
      ${'(32)9222-22835'},
      ${' 32922222835'},
      ${'32922222835 '},
    `('rejects invalid phone numbers.', ({ invalidPhone }) => {
      expect(() => new PhoneNumber(invalidPhone)).toThrow(PhoneNumberValidationError)
    });

    it.each`
      validPhone
      ${32984092323},
      ${'32984092323'},
      ${'11984092323'},
      ${'555'},
    `('accepts valid phone numbers.', ({ validPhone }) => {
      expect(() => new PhoneNumber(validPhone)).not.toThrow();
    });

    it('is JSON serializable.', () => {
      const number = '32988128811';
      const pn = new PhoneNumber(number);

      const json = JSON.stringify({ phoneNumber: pn });

      expect(JSON.parse(json)).toEqual(expect.objectContaining({
        phoneNumber: number,
      }));
    });

    it('reports its size.', () => {
      const pn = new PhoneNumber('1234567891234567890');

      expect(pn.size).toBe(19);
    });
  });

  describe('Message', () => {

    it.each`
      validMessage
      ${TEXT_160_CHARS},
      ${'Oi, Joao!\n\rQuer ficar milhonario???'},
      ${'Que a forca esteja com voce!'},
      ${'Codigo de verificacao: 523532\n(Valido ate 25-11-2019)'},
    `('accepts valid messages.', ({ validMessage }) => {
      expect(() => new Message(validMessage)).not.toThrow();
    });

    it('is JSON serializable.', () => {
      const text = 'Parabens! Voce foi sorteado...'
      const msg = new Message(text);

      const json = JSON.stringify({ message: msg });

      expect(JSON.parse(json)).toEqual(expect.objectContaining({
        message: text,
      }));
    });

    it('rejects a message that exceeds 160 characters.', () => {
      
      expect.assertions(3);

      // Sanity check.
      expect(TEXT_161_CHARS.length).toBe(161);

      try {
        new Message(TEXT_161_CHARS);
      } catch (err) {
        expect(err).toBeInstanceOf(MessageValidationError);
        expect(err.code).toBe('MESSAGE_TOO_LONG_ERR');
      }
    });

    it('informs its size', () => {
      const msg = new Message('Codigo de verificacao: 523532\n(Valido ate 25-11-2019)');

      expect(msg.size).toBe(53);
    });
  });

  describe('SmsClient', () => {

    it('sends a SMS', async () => {
      const requestMock = jest.fn();

      const client = new SmsClient({
        request: requestMock,
      });

      const data = ({
        celular: '32988125723',
        msg: 'Aviso: seus creditos expiram em 12-05-20',
      });

      await client.sendSms(data.celular, data.msg);

      expect(requestMock).toHaveBeenCalledTimes(1);
    });

    it('throws an error if http request fails.', () => {
      const requestMock = jest.fn(() => {
        return Promise.reject(new Error('Http request error'));
      });

      const client = new SmsClient({
        request: requestMock,
      });

      const data = ({
        celular: '32988125723',
        msg: 'Aviso: seus creditos expiram em 12-05-20',
      });

      return expect(client.sendSms(data.celular, data.msg))
        .rejects
        .toThrow(HttpRequestError);
    });
  });
});


src/errors.js
class SmsClientError extends Error {

  constructor(e) {
    super(e);

    this.message = e.message || e;
    this.extensions = {};

    Object.defineProperty(this, 'name', { value: this.constructor.name });

    Error.captureStackTrace(this, this.constructor);
  }
}

class HttpRequestError extends SmsClientError {
  constructor(e) {
    super(e);

    this.extensions = {
      originalError: e,
    };
  }
}

class ValidationError extends SmsClientError {
  // No implementation
}

class PhoneNumberValidationError extends ValidationError {

  constructor(e, { invalidPhoneNumber }) {
    super(e);

    this.extensions = {
      invalidPhoneNumber,
    };
  }
}

class MessageValidationError extends ValidationError {

  constructor(e, code, { invalidMessage }) {
    super(e);

    this.code = code;
    this.extensions = {
      invalidMessage,
    };
  }
}

module.exports = {
  SmsClientError,
  ValidationError,
  PhoneNumberValidationError,
  MessageValidationError,
  HttpRequestError,
};


src/client.js
const {
  MessageValidationError,
  PhoneNumberValidationError,
  HttpRequestError,
} = require('./errors');

const phoneNumberRegex = new RegExp(`^[0-9]+\$`);

class Message {

  get size() {
    return this._msg.length;
  }

  constructor(messageStr) {
    this._validate(messageStr);
    this._msg = messageStr;
  }

  _validate(messageStr) {
    if (messageStr.length > 160) {
      throw new MessageValidationError(
        'Message overseeds the maximum of 140 characters',
        'MESSAGE_TOO_LONG_ERR',
        { invalidMessage: messageStr }
      );
    }
  }

  valueOf() {
    return this._msg;
  }

  toString() {
    return this.valueOf();
  }

  toJSON() {
    return this.toString();
  }
}

class PhoneNumber {

  get size() {
    return this._phoneNumber.length;
  }

  constructor(phoneNumberStr) {
    this._validate(phoneNumberStr);
    this._phoneNumber = phoneNumberStr;
  }

  _validate(phoneNumberStr) {
    if (!phoneNumberRegex.test(phoneNumberStr)) {
      throw new PhoneNumberValidationError(
        `Phone number ${phoneNumberStr} is invalid`,
        { invalidPhoneNumber: phoneNumberStr }
      );
    }
  }

  valueOf() {
    return this._phoneNumber;
  }

  toString() {
    return this.valueOf();
  }

  toJSON() {
    return this.toString();
  }
}

class SmsClient {

  constructor({ authToken, request, url }) {
    this._request = request;
    this._url = url;
    this._method = 'post';
    this._headers = {
      'Content-Type': 'application/json',
      'token_auth': authToken,
    };
  }

  async sendSms(phoneNumber, message) {
    const data = JSON.stringify({
      celular: new PhoneNumber(phoneNumber),
      msg: new Message(message),
    });

    let response;

    try {
      response = await this._request({
        data,
        url: this._url,
        method: this._method,
        headers: this._headers,
      });
    } catch (err) {
      throw new HttpRequestError(err);
    }

    return response;
  }
}

module.exports = {
  SmsClient,
  PhoneNumber,
  Message,
};
